### ISSUE TITLE
ogm bug: first call to `.add` yields a protocol error between two previously "pushed" nodes

### ISSUE THREAD

**Issue by wgwz** [technige/py2neo#604]
TL;DR: unhelpful/misleading/unexplained error message

here's the relatively self-contained mcve that reproduces the issue:

```
from py2neo.ogm import GraphObject, Property, RelatedTo, RelatedFrom
from uuid import uuid4
from myapp.extensions import graph
class Activity(GraphObject):
    arbitrary = Property()
    def __init__(self):
        self.arbitrary = 'test'

class User(GraphObject):
    __primarykey__ = 'email'
    email = Property()
    uid = Property()
    name = Property()
    has_activity = RelatedTo(Activity)
    def __init__(self, name, email):
        self.uid = str(uuid4())
        self.name = name
        self.email = email

user = User(name='tester', email='tester@domain')
acti = Activity()
graph.push(user)
graph.push(acti)
user.has_activity.add(acti)
```
yields:
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/py2neo/ogm.py", line 171, in add
    related_objects = self._related_objects
  File "/usr/local/lib/python2.7/site-packages/py2neo/ogm.py", line 161, in _related_objects
    self.__db_pull__(remote_node.graph)
  File "/usr/local/lib/python2.7/site-packages/py2neo/ogm.py", line 228, in __db_pull__
    for r in graph.match(*self.__match_args):
  File "/usr/local/lib/python2.7/site-packages/py2neo/database/__init__.py", line 610, in match
    cursor = self.run(" ".join(clauses), parameters)
  File "/usr/local/lib/python2.7/site-packages/py2neo/database/__init__.py", line 731, in run
    return self.begin(autocommit=True).run(statement, parameters, **kwparameters)
  File "/usr/local/lib/python2.7/site-packages/py2neo/database/__init__.py", line 370, in begin
    return self.transaction_class(self, autocommit)
  File "/usr/local/lib/python2.7/site-packages/py2neo/database/__init__.py", line 1249, in __init__
    self.session = driver.session()
  File "/usr/local/lib/python2.7/site-packages/py2neo/packages/neo4j/v1/session.py", line 126, in session
    connection = connect(self.address, self.ssl_context, **self.config)
  File "/usr/local/lib/python2.7/site-packages/py2neo/packages/neo4j/v1/bolt.py", line 482, in connect
    return Connection(s, der_encoded_server_certificate=der_encoded_server_certificate, **config)
  File "/usr/local/lib/python2.7/site-packages/py2neo/packages/neo4j/v1/bolt.py", line 241, in __init__
    self.fetch()
  File "/usr/local/lib/python2.7/site-packages/py2neo/packages/neo4j/v1/bolt.py", line 344, in fetch
    handler(*fields)
  File "/usr/local/lib/python2.7/site-packages/py2neo/packages/neo4j/v1/bolt.py", line 233, in on_failure
    raise ProtocolError(metadata.get("message", "INIT failed"))
ProtocolError: Missing username and password
```
but running this immediately after the traceback occurs, this works:
```
user.has_activity.add(acti)
graph.push(user)  
```

- why the protocol error in this first example? 
- also curious is that the second call to `user.has_activity.add(acti)` works. 
- after the final `graph.push(user)` i have three nodes in the db:
    + one presumably stale Activity node
    + a User node and an Activity node with the HAS_ACTIVITY relationship between

here's an mcve that doesn't have this issue and correctly produces two nodes with a HAS_ACTIVITY relationship between them:

```
from py2neo.ogm import GraphObject, Property, RelatedTo, RelatedFrom
from uuid import uuid4
from myapp.extensions import graph

class Activity(GraphObject):
   arbitrary = Property()
   def __init__(self):
       self.arbitrary = 'test'

class User(GraphObject):
    __primarykey__ = 'email'
    email = Property()
    uid = Property()
    name = Property()
    has_activity = RelatedTo(Activity)
    def __init__(self, name, email):
        self.uid = str(uuid4())
        self.name = name
        self.email = email

user = User(name='tester', email='tester@domain')
acti = Activity()
user.has_activity.add(acti)
graph.push(user)
```

i am connecting to the database with:

```
params = {
        'host': app.config['GRAPH_HOST'],
        'port': app.config['GRAPH_PORT'],
        'user': app.config['GRAPH_USER'],
        'password': app.config['GRAPH_PASSWORD'],
        'bolt': app.config['GRAPH_BOLT_ENABLED']
    }
graph = Graph(**params)
```

`py2neo==3.1.2`
python 2.7.12
